source/Calamari.Common/Features/StructuredVariables/PropertiesFormatVariableReplacer.cs

```diff
@@ -0,0 +1,82 @@
+﻿using System;
+using System.Text.RegularExpressions;
+using Calamari.Common.Plumbing.Extensions;
+using Calamari.Common.Plumbing.FileSystem;
+using Calamari.Common.Plumbing.Logging;
+using Calamari.Common.Plumbing.Variables;
+using JavaPropertiesParser;
+using JavaPropertiesParser.Expressions;
+using JavaPropertiesParser.Utils;
+
+namespace Calamari.Common.Features.StructuredVariables
+{
+    public class PropertiesFormatVariableReplacer : IFileFormatVariableReplacer
+    {
+        readonly ICalamariFileSystem fileSystem;
+        readonly ILog log;
+
+        public PropertiesFormatVariableReplacer(ICalamariFileSystem fileSystem, ILog log)
+        {
+            this.fileSystem = fileSystem;
+            this.log = log;
+        }
+        
+        readonly Regex octopusReservedVariablePattern = new Regex(@"^Octopus([^:]|$)", RegexOptions.IgnoreCase | RegexOptions.Compiled);
+
+        public string FileFormatName => StructuredConfigVariablesFileFormats.Properties;
+
+        public bool IsBestReplacerForFileName(string fileName)
+        {
+            return fileName.EndsWith(".properties", StringComparison.InvariantCultureIgnoreCase);
+        }
+
+        public void ModifyFile(string filePath, IVariables variables)
+        {
+            try
+            {
+                var (fileText, encoding) = EncodingDetectingFileReader.ReadToEnd(filePath);
+
+                var parsed = Parser.Parse(fileText);
+                var updated = parsed.Mutate(expr => TryReplaceValue(expr, variables));
```

Comment:
Treating ASTs as immutable is generally the way to go. Our parser tries to treat them as immutable, but .net 4.0 doesn't have `IReadOnlyList<T>` or `ImmutableList`, so we just model the top level expressions in an array. While we could reach into the document and just change things, I've tried to keep this consistent with how I'd like the parser to evolve in future, once we don't have to support .net 4.0.

Reply:
I much prefer immutability. Much safer.